types/wlr_output.c

```diff
@@ -288,170 +368,107 @@ static void handle_cursor_surface_commit(struct wl_listener *listener,
-		cursor.surface_destroy);
-
-	wl_list_remove(&output->cursor.surface_commit.link);
-	wl_list_remove(&output->cursor.surface_destroy.link);
-	output->cursor.surface = NULL;
+	struct wlr_output_cursor *cursor = wl_container_of(listener, cursor,
+		surface_destroy);
+	output_cursor_reset(cursor);
 }
 
-void wlr_output_set_cursor_surface(struct wlr_output *output,
+void wlr_output_cursor_set_surface(struct wlr_output_cursor *cursor,
 		struct wlr_surface *surface, int32_t hotspot_x, int32_t hotspot_y) {
 	if (surface && strcmp(surface->role, "wl_pointer-cursor") != 0) {
 		return;
 	}
 
-	output->cursor.hotspot_x = hotspot_x;
-	output->cursor.hotspot_y = hotspot_y;
+	if (surface) {
+		cursor->width = surface->current->width;
+		cursor->height = surface->current->height;
+	}
+	cursor->hotspot_x = hotspot_x;
+	cursor->hotspot_y = hotspot_y;
 
-	if (surface && surface == output->cursor.surface) {
-		if (output->impl->set_cursor && !output->cursor.is_sw) {
-			// Only update the hotspot
-			output->impl->set_cursor(output, NULL, 0, 0, 0, hotspot_x,
-				hotspot_y, false);
+	if (surface && surface == cursor->surface) {
+		if (cursor->output->hardware_cursor == cursor &&
+				cursor->output->impl->set_cursor) {
+			// If the surface hasn't changed and it's an hardware cursor, only
+			// update the hotspot
+			cursor->output->impl->set_cursor(cursor->output, NULL, 0, 0, 0,
+				hotspot_x, hotspot_y, false);
 		}
 		return;
 	}
 
-	if (output->cursor.surface) {
-		wl_list_remove(&output->cursor.surface_commit.link);
-		wl_list_remove(&output->cursor.surface_destroy.link);
-		output->cursor.surface = NULL;
-	}
+	output_cursor_reset(cursor);
 
-	// Disable hardware cursor
+	// Disable hardware cursor for surfaces
 	// TODO: support hardware cursors
-	output->cursor.is_sw = true;
-	if (output->impl->set_cursor) {
-		output->impl->set_cursor(output, NULL, 0, 0, 0, hotspot_x, hotspot_y,
+	if (cursor->output->hardware_cursor == cursor &&
+			cursor->output->impl->set_cursor) {
+		cursor->output->impl->set_cursor(cursor->output, NULL, 0, 0, 0, 0, 0,
 			true);
+		cursor->output->hardware_cursor = NULL;
 	}
 
-	//output->cursor.is_sw = output->impl->set_cursor == NULL;
-	output->cursor.surface = surface;
+	cursor->surface = surface;
 
 	if (surface != NULL) {
-		wl_signal_add(&surface->events.commit, &output->cursor.surface_commit);
-		wl_signal_add(&surface->events.destroy,
-			&output->cursor.surface_destroy);
-		commit_cursor_surface(output, surface);
+		wl_signal_add(&surface->events.commit, &cursor->surface_commit);
+		wl_signal_add(&surface->events.destroy, &cursor->surface_destroy);
+		output_cursor_commit(cursor);
 	} else {
-		set_cursor(output, NULL, 0, 0, 0, hotspot_x, hotspot_y);
+		// TODO: if hardware cursor, disable cursor
 	}
 }
 
-bool wlr_output_move_cursor(struct wlr_output *output, int x, int y) {
-	output->cursor.x = x;
-	output->cursor.y = y;
+bool wlr_output_cursor_move(struct wlr_output_cursor *cursor, int x, int y) {
```

Comment:
wlr-cursor coordinates are doubles but these are int so there is some rounding going on here and output-cursor coordinates will usually be different from cursor coordinates.

Reply:
Is that a problem?

Reply:
The worst that can happen is that the cursor will be rendered 1px away from where it really is. If we can render the cursor at double coordinates, we absolutely should. If not, that's fine because the wlr-cursor will still have the right coordinates for the next move so the inaccuracies will not accumulate as the mouse moves.

Reply:
Hardware cursors only support integers, so I don't think that's possible.